### PR TITLE
Add runtime command flow foundation

### PR DESCRIPTION
**Pull request by Vasanthdev2004**
## Summary
- add a Zero runtime context that resolves provider/model/profile/autonomy/tool filters before agent runs
- add TUI command registry and wire /help, /doctor, /config, /search, /context, /permissions, and /clear
- add intent-based model profiles: fast, balanced, deep, cheap
- expand zero exec with --profile, --reasoning-effort, --auto, --enabled-tools, --disabled-tools, --list-tools, and --max-turns
- keep gated tools behind TUI approval flow and let agent loop enforce enabled/disabled tool filters
- remove startup shortcut hint clutter from the prior UI state

## Validation
- bun run typecheck
- bun test ./tests --timeout 15000
- bun run build
- bun run smoke:build

## Notes
- Clean-room implementation: D was used only to identify product-flow patterns, not copied code.
- App.tsx was already above the preferred 500-line target; this PR keeps the behavior focused, and the next cleanup should extract TUI command handlers into a smaller controller module.

### CODE REVIEW

File: src/tui/App.tsx [hunk 1]
```diff
@@ -453,6 +471,23 @@ export const App: React.FC = () => {
     setMessages((prev) => [...prev, { type: 'system', content }]);
   };
 
+  const requestToolApproval = async (request: ToolApprovalRequest): Promise<ToolApprovalDecision> => {
+    if (approvalGrantsRef.current.has(request.grantKey)) {
+      return 'allow';
+    }
+
+    setPendingApproval(request);
+    return new Promise<ToolApprovalDecision>((resolve) => {
```
Comment (gnanam1990) [inline src/tui/App.tsx:578]: `requestToolApproval` stores a single resolver in `approvalResolverRef.current`. Because the agent loop executes all tool calls for a turn with `Promise.all`, a provider response containing two prompt-gated tool calls can invoke this callback twice before the user responds. The second call overwrites the first resolver, so the first approval promise is orphaned and `runAgent` can hang forever. Please serialize prompt-gated tool execution or queue approval requests so each pending approval keeps its own resolver.

**Review by gnanam1990 — CHANGES_REQUESTED** [review]
## Automated Hermes Agent rereview

Verdict: **Changes requested**

This PR has not previously been reviewed by me, so I inspected the full current diff against `origin/main`.

### Existing review/comment context considered
- Existing PR reviews: none.
- Existing PR comments: none.

### What changed in this PR
- Adds `ask` permission mode in the agent loop and advertises prompt-gated tools only in ask/unsafe modes.
- Adds TUI state and an approval panel for bash/file/patch actions with y/n/always-this-session controls.
- Adds Bun tests for advertising behavior, deny/allow/session decisions, and approval-panel rendering.

### Blocking finding
- The TUI stores only one pending approval resolver. If a provider returns multiple prompt-gated tool calls in the same assistant response, `runAgent` starts approvals through `Promise.all`, so `requestToolApproval` can be called multiple times before the user answers. Each call overwrites `approvalResolverRef.current`; the first approval promise can no longer be resolved, leaving the agent loop hung and potentially showing only the last approval request. Please serialize approval-gated tool execution, queue pending approval requests in the TUI, or otherwise ensure each prompt-gated tool call has its own resolver lifecycle before merging.

### Validation
- `npx --yes bun install --frozen-lockfile` ✅
- `npx --yes bun test ./tests --timeout 15000` ✅ (200 pass, 0 fail)
- `npx --yes bun run typecheck` ✅
- `npx --yes bun run build` ✅
- `npx --yes bun run smoke:build` ✅

Once the concurrent-approval deadlock is fixed, this should be close; the safety split between auto/ask/unsafe and the test coverage for single approvals look good.

---
*Automated review by Hermes Agent*

**Comment by gnanam1990** [comment]
@vasanthdev2004 @anandh8x please review this PR.

**Comment by Vasanthdev2004** [comment]
### Blockers

- `src/tui/App.tsx:474` stores only one pending approval resolver, while `src/agent/loop.ts:207` executes all assistant tool calls through `Promise.all`. If a provider emits two prompt-gated tool calls in one assistant turn, `requestToolApproval` is called twice before the user answers; the second call overwrites `approvalResolverRef.current`, so the first approval promise can never resolve and the agent loop hangs. Please serialize prompt-gated tool execution or queue approval requests so each pending tool call keeps its own resolver lifecycle.

### Non-Blocking

- None.

### Looks Good

- The auto/ask/unsafe split now correctly keeps deny tools out of the advertised tool list and requires an explicit grant before prompt-gated tools reach the registry.
- The TUI approval panel gives useful previews for bash, file writes, edit_file, and apply_patch actions.
- 371 additions, 11 deletions — focused feature surface with targeted tests for allow, deny, allow-session, and render behavior.
- Validation: `npx --yes bun install --frozen-lockfile`, `bun test ./tests --timeout 15000` (200/200 passing), `bun run typecheck`, `bun run build`, `bun run smoke:build`, and `git diff --check` passed. Additional concurrent-approval probe reproduced the hang: two approval requests fired, resolving the single visible resolver left the run stuck.

---

**Verdict: Changes Requested** — Fix the multi-tool approval deadlock before merging.

**Review by anandh8x — APPROVED** [review]
## Review: TUI tool approval flow

**No blockers.** Clean integration of ask-mode permission handling with a focused TUI approval panel.

### What's good
- **Three permission modes** — `auto` (existing, allow-only tools), `ask` (prompts for prompt-gated tools), `unsafe` (existing, grants everything). Clear separation of concerns
- **Grant key** — `{permission}:{sideEffect}` string scopes session-level auto-approvals by safety class. Cleaner than tracking by individual tool name
- **`onToolApproval` callback** — Returns `Promise<ToolApprovalDecision>`, allowing both sync and async handlers. The TUI uses a promise-based resolver pattern (via `approvalResolverRef`)
- **Dual grant check** — TUI checks `approvalGrantsRef` before showing the prompt (avoids flickering already-granted approvals), loop checks its own `approvalGrants` Set for the actual permission flag. Both are consistent
- **Per-tool previews** — `bash` shows command+cwd, `write_file` shows path+bytes+overwrite, `edit_file` shows path+delta+replace_all, `apply_patch` parses diff for files+line counts. Only relevant information, no noise
- **Key bindings** — y=allow, n=deny, a=allow-session. Blocks other input while pending. The panel shows all three clearly with colored brackets
- **Test coverage** — 5 new tests covering: auto mode filtering, unsafe mode, ask mode advertising, deny (file not written), allow (file written), allow-session (skips repeated prompt), and delta-before-start edge case
- **TUI test** — Renders pending approval panel with bash preview, verifies [y]/[n]/[a] are present

### Minor observations
- **Grant key scope** — `prompt:shell` or `prompt:filesystem` means approving bash auto-approves all shell tools, and approving `write_file` auto-approves `edit_file` in the same class by default. Intended (grouping by safety class), but users might expect per-tool granularity
- **Test artifact** — `tests/agent-loop.test.ts` line 30 contains `cwd: 'D:\\codings\\Opensource\\Zero'` (contributor local path). Not a blocker — the test uses `pendingApproval.parsedArgs.cwd` which is just display text
- **`approvalGrants` lifespan** — Created per `runAgent` call, never cleared mid-loop. Correct — session = one agent invocation. Survives multiple turns

**Comment by Vasanthdev2004** [comment]
### Blockers

- The prior multi-tool approval deadlock is still present. `src/agent/loop.ts:207` executes all assistant tool calls through `Promise.all`, while `src/tui/App.tsx:474` still stores a single `approvalResolverRef`. A provider that emits two prompt-gated calls in one turn creates two approval requests before user input; the later request overwrites the earlier resolver and the agent run can hang. The new commit only removes startup shortcut hints, so it does not address the blocker.

### Non-Blocking

- None.

### Looks Good

- The shortcut-hint cleanup is harmless and keeps the TUI surface less noisy.
- The single-approval tests and TUI approval-panel render coverage still pass.
- 379 additions, 46 deletions — focused approval-flow surface plus UI cleanup.
- Validation: `npx --yes bun install --frozen-lockfile`, targeted agent/TUI tests (16/16 passing), `bun run typecheck`, full `bun test ./tests --timeout 15000` (200/200 passing), `bun run build`, `bun run smoke:build`, and `git diff --check` passed. Additional concurrent-approval probe still reproduced the hang: two approval requests fired, resolving the latest left the run timed out.

---

**Verdict: Changes Requested** — Please serialize/queue prompt-gated approvals so each pending tool call has its own resolver lifecycle.

**Review by gnanam1990 — CHANGES_REQUESTED** [review]
## Automated Hermes Agent rereview

Verdict: **Changes requested**

This rereview was triggered because the latest PR commit (`ac549aa`, 2026-06-03T09:27:28Z) is newer than my prior review on `497bac0` (2026-06-03T09:21:05Z).

### What changed since my previous review
- `ac549aa Remove startup shortcut hints` removes startup shortcut hints and updates TUI render expectations.
- The delta only touched `src/tui/CommandSuggestions.tsx`, `src/tui/StartupScreen.tsx`, `src/tui/TuiPromptBox.tsx`, `src/tui/startup/ShortcutHints.tsx`, and `tests/tui-shell-render.test.ts`.

### Existing reviews/comments considered
- Considered @anandh8x's approval noting the TUI approval panel integration and tests.
- Considered the PR comments from @Vasanthdev2004 calling out that the multi-tool approval deadlock remains.
- Rechecked my prior blocker against the current head.

### Blocking issue
The prior multi-tool approval blocker is still present. `src/agent/loop.ts` still executes assistant tool calls concurrently with `Promise.all`, while `src/tui/App.tsx` still keeps only one pending approval request/resolver (`pendingApproval` plus `approvalResolverRef`). If a model emits two prompt-gated tool calls in the same assistant turn, the second `requestToolApproval()` overwrites the first resolver and visible pending request. The first promise can remain unresolved, so `Promise.all(toolPromises)` can hang indefinitely. This is a correctness blocker for ask-mode tool approval.

Please serialize prompt-gated approvals/tool execution, or maintain a queue/map of pending approval resolvers keyed by tool call id so concurrent tool calls cannot overwrite each other. Add a regression test with two prompt-gated tool calls in one assistant turn.

### Validation
Passed locally on current PR head using `npx --yes bun`:
- `bun install --frozen-lockfile`
- `bun test` — 200 pass / 0 fail
- `bun run typecheck`
- `bun run build`

The validation passing does not cover the concurrent approval deadlock above; no regression test currently exercises it.

**Review by gnanam1990 — CHANGES_REQUESTED** [review]
## Review update

Verdict: **Changes requested**

I reviewed the latest head `ac549aa` after the previous changes-requested review on `497bac0`. The only new commit removes the startup shortcut hints (`Enter`, `Tab`, `Ctrl+C`) and updates the related render assertions. I also considered the existing review/comment thread, especially the blocker around multiple prompt-gated tool calls in one assistant turn.

The earlier blocker is still present:

- `src/agent/loop.ts` still executes all assistant tool calls concurrently with `Promise.all`.
- `src/tui/App.tsx` still stores only one pending approval resolver in `approvalResolverRef.current`.
- If the model emits two prompt-gated tool calls in the same turn, both approval requests can be created before the user answers. The second request overwrites the first resolver, leaving the first promise unresolved and the agent run can hang.

Please serialize prompt-gated approval/execution or add a real approval queue so each pending tool call keeps its own resolver and is resolved in order.

Validation I ran locally:

- `bun install --frozen-lockfile` passed
- `bun test ./tests/agent-loop.test.ts ./tests/tui-shell-render.test.ts --timeout 15000` passed — 16 tests
- `bun x tsc --noEmit` passed
- `bun test ./tests --timeout 15000` passed — 200 tests
- `bun run build` passed
- `bun run smoke:build` passed
- `./zero --help` passed

GitHub checks are also green on this head, and the PR merges cleanly into `origin/main`. The remaining blocker is the unresolved multi-tool approval deadlock above.

**Review by gnanam1990 — CHANGES_REQUESTED** [review]
## Review update

Verdict: **Changes requested**

I reviewed the latest head `d0c4e81` after my previous review on `ac549aa`. The new commit adds the runtime command flow foundation: shared run context creation, model profiles/reasoning/autonomy/tool filters, TUI command registry additions, and related CLI/TUI/runtime tests.

I also rechecked the existing review/comment thread. The prior blocker around multiple prompt-gated tool calls is still unresolved:

- `src/agent/loop.ts` still executes assistant tool calls concurrently via `Promise.all`.
- `src/tui/App.tsx` still keeps a single pending approval resolver in `approvalResolverRef.current`.
- If a provider returns two prompt-gated tool calls in one assistant turn, the second `requestToolApproval` can overwrite the first resolver before the user responds. That leaves the first approval promise unresolved and can hang the agent loop.

Please serialize prompt-gated approvals/tool execution, or implement an approval queue where each pending tool call owns its resolver lifecycle and is resolved in order.

Validation I ran locally:

- `bun install --frozen-lockfile` passed
- Targeted tests passed: `tests/agent-loop.test.ts`, `tests/tui-shell-render.test.ts`, `tests/tui-command-registry.test.ts`, `tests/zero-runtime.test.ts`, `tests/headless-exec.test.ts` — 31 tests
- `bun x tsc --noEmit` passed
- `bun test ./tests --timeout 15000` passed — 209 tests
- `bun run build` passed
- `bun run smoke:build` passed
- `./zero --help` passed

GitHub checks are green, `git diff --check` is clean, and the PR merges cleanly into `origin/main`. The only blocker I see is the unresolved multi-tool approval deadlock above.

**Comment by Vasanthdev2004** [comment]
### Blockers

- The multi-tool approval deadlock is still present at `src/agent/loop.ts:273` plus `src/tui/App.tsx:572-580`. `runAgent` still starts all tool executions with `Promise.all`, and the TUI still stores a single `approvalResolverRef`; a model turn with two prompt-gated calls can create two approval promises before user input, the second resolver overwrites the first, and the run can hang. I reproduced the current head with a two-`bash` approval probe: two approvals fired, resolving the visible/latest resolver left the agent timed out. Please serialize prompt-gated execution or queue/map pending approvals per tool call.

### Non-Blocking

- None.

### Looks Good

- The runtime command/context foundation and tool/model command helpers are wired coherently, with useful tests for autonomy mapping, tool filters, model profiles, and TUI command registry behavior.
- The new focused suites pass alongside the existing full suite.
- 1,340 additions, 122 deletions — broader runtime-command foundation plus the existing approval-flow surface.
- Validation: `npx --yes bun install --frozen-lockfile`, targeted agent/TUI/runtime/headless/command/model tests (36/36 passing), `bun run typecheck`, full `bun test ./tests --timeout 15000` (209/209 passing), `bun run build`, `bun run smoke:build`, `git diff --check origin/main..HEAD`, and `git diff --check` passed. The additional concurrent approval probe failed as described above.

---

**Verdict: Changes Requested** — The current head still needs serialized or queued prompt-gated approvals before merging.

**Comment by Vasanthdev2004** [comment]
### Blockers

- The multi-tool approval deadlock is still present at `src/agent/loop.ts:217-273` plus `src/tui/App.tsx:572-580`. `runAgent` still starts all assistant tool executions concurrently with `Promise.all`, and the TUI still stores only one `approvalResolverRef`; when a provider emits two prompt-gated tool calls in one turn, both approvals are requested before user input and the second resolver overwrites the first. I reproduced this on the current head: two `bash` approval requests fired, resolving the visible/latest resolver left the agent timed out. Please serialize prompt-gated tool execution or implement a real approval queue/map so each pending call owns its resolver lifecycle.

### Non-Blocking

- None.

### Looks Good

- The force-pushed current head is now mergeable with `origin/main`, and the runtime command flow, profile selection, tool filters, and command-registry wiring are still cohesive.
- Targeted runtime/TUI/headless coverage and the full test suite pass on this branch.
- 1,377 additions, 129 deletions — broad runtime-command foundation plus TUI approval flow and focused tests.
- Validation: `npx --yes bun install --frozen-lockfile`, targeted agent/TUI/runtime/headless tests (36/36 passing), `npx --yes bun run typecheck`, full `npx --yes bun test ./tests --timeout 15000` (219/219 passing), `npx --yes bun run build`, `npx --yes bun run smoke:build`, `./zero --help`, `git diff --check origin/main..HEAD`, and `git diff --check` passed. The concurrent approval probe failed as described above.

---

**Verdict: Changes Requested** — The PR still needs serialized or queued prompt-gated approvals before it can safely merge.

**Review by gnanam1990 — CHANGES_REQUESTED** [review]
## Review update

Verdict: **Changes requested**

I reviewed the latest head `4e1e09f` after my previous review on `d0c4e81`. The branch was force-pushed; the current stack now includes the approval-flow commits rebased on newer `main`, plus the stream-json protocol/foundation work from `main`.

I also rechecked the existing review/comment thread, including the repeated blocker around multiple prompt-gated tool calls. That blocker is still present on this head:

- `src/agent/loop.ts` still starts all assistant tool executions concurrently and awaits `Promise.all`.
- `src/tui/App.tsx` still keeps a single pending approval resolver in `approvalResolverRef.current`.
- If one assistant turn contains two prompt-gated tool calls, two approval requests can be created before the user answers. The second request can overwrite the first resolver, leaving the first promise unresolved and hanging the agent loop.

Please serialize prompt-gated tool approval/execution, or implement a queue so each pending approval has its own resolver and is resolved in order.

Validation I ran locally:

- `bun install --frozen-lockfile` passed
- Targeted tests passed: `tests/agent-loop.test.ts`, `tests/tui-shell-render.test.ts`, `tests/tui-command-registry.test.ts`, `tests/zero-runtime.test.ts`, `tests/headless-exec.test.ts`, `tests/zero-stream-json.test.ts` — 41 tests
- `bun x tsc --noEmit` passed
- `bun test ./tests --timeout 15000` passed — 219 tests
- `bun run build` passed
- `bun run smoke:build` passed
- `./zero --help` passed

GitHub checks are green, `git diff --check` is clean, and the PR currently merges cleanly into `origin/main`. The remaining blocker is still the multi-tool approval deadlock.

**Comment by gnanam1990** [comment]
Yes, changes were made on the branch, but the latest push does not resolve the blocking issue from the review.

Latest reviewed head: `4e1e09f`

The branch added runtime command flow work and picked up newer main changes, but the multi-tool approval deadlock is still present:

- `runAgent` can still dispatch multiple tool calls concurrently with `Promise.all`
- the TUI still tracks only one pending approval resolver via `approvalResolverRef.current`
- if two prompt-gated tool calls arrive in one assistant turn, the second approval can overwrite the first resolver, leaving the first tool call unresolved

So the PR still needs changes before approval. CI being green is good, but it does not cover this interactive approval path.

**Comment by Vasanthdev2004** [comment]
### Blockers

None found.

### Non-Blocking

- None.

### Looks Good

- The prior multi-tool approval deadlock is fixed: `runAgent` now serializes prompt-gated tool execution, so the TUI's single pending resolver is not overwritten by concurrent approvals.
- Permission behavior still fails closed in auto mode, keeps prompt-gated tools behind ask/unsafe modes, and supports allow-session grants.
- Validation passed: 220/220 tests, typecheck, build, smoke:build, git diff checks, and CI smoke checks on Ubuntu/macOS/Windows/Performance are green.
- Targeted approval probe confirmed two bash calls in one assistant turn request approvals one at a time and resolve successfully; 1,483 additions, 158 deletions covers the runtime command/TUI foundation.

---

**Verdict: Approve** — Clean runtime command flow foundation; self-authored PR so this is a review comment only.

**Review by gnanam1990 — APPROVED** [review]
## Automated Hermes Agent rereview

Verdict: **Approved**

I rereviewed the new head `a94abf4` after the previous changes-requested review on `4e1e09f`.

### Resolution checked
- The previous blocker was the prompt-gated approval deadlock: `runAgent` executed same-turn tool calls concurrently with `Promise.all`, while the TUI held a single pending approval resolver.
- The new commit serializes assistant tool execution in `src/agent/loop.ts` (`for ... await executeToolCall(...)`), so an interactive approval request must resolve before the next prompt-gated tool call can request approval.
- The TUI still has a single visible pending resolver, but that is now compatible with the serialized agent loop.
- A regression test now covers two prompt-gated `bash` calls in one assistant turn and verifies only the first approval appears until it is resolved, then the second approval proceeds and both tool results return in order.

### Validation run locally
- `npx --yes bun test ./tests/agent-loop.test.ts ./tests/tui-shell-render.test.ts ./tests/zero-runtime.test.ts --timeout 15000` ✅ — 22 pass / 0 fail
- `npx --yes tsc --noEmit` ✅

GitHub checks are also green on this head. I do not see remaining blockers from my prior feedback.

---
*Automated rereview by Hermes Agent*

**Review by gnanam1990 — APPROVED** [review]
## Review update

Verdict: **Approved**

Reviewed PR #30 again at head `a94abf4d18d45fc6371e23b15b5abb116aedd674`.

Existing review/comment context considered:
- Prior blocker: multiple prompt-gated tool calls could overlap because tool execution used `Promise.all` while the TUI tracked only one pending approval resolver.
- Latest author/comment context says the blocker is resolved by serializing prompt-gated approvals.

What I checked:
- `runAgent` now executes assistant tool calls sequentially, so overlapping approvals cannot overwrite the single pending approval UI state.
- Prompt-gated tools are still hidden in auto mode, advertised in ask/unsafe mode, and deny tools remain excluded.
- The new agent-loop test covers serialized multiple prompt-gated approvals from the same assistant turn.

Validation:
- `bun install --frozen-lockfile` passed
- `bun test` passed: 220 tests, 0 failures
- `bun run typecheck` passed
- `bun run build` passed
- GitHub checks are passing: Smoke on Ubuntu/macOS/Windows and Performance Smoke

No blockers found.
